EventFilter/EcalRawToDigi/plugins/alpaka/UnpackPortable.dev.cc

```diff
@@ -0,0 +1,457 @@
+              const bool regular = (chStatus == CH_DISABLED || chStatus == CH_SUPPRESS);
+              const bool problematic =
+                  (chStatus == CH_TIMEOUT || chStatus == CH_HEADERERR || chStatus == CH_LINKERR ||
+                   chStatus == CH_LENGTHERR || chStatus == CH_IFIFOFULL || chStatus == CH_L1AIFIFOFULL);
+              if (!(regular || problematic)) {
+                exp_ttids[nCh] = ch;
+                ++nCh;
+              }
+            }
+          }
+
+          //
+          // print Tower block headers
+          //
+          uint8_t ntccblockwords = isBarrel ? 18 : 36;
+          auto const* tower_blocks_start = buffer + 9 + ntccblockwords + 6;
+          auto const* trailer = buffer + (size / 8 - 1);
+          auto const* current_tower_block = tower_blocks_start;
+          uint8_t iCh = 0;
+          uint8_t next_tower_id = exp_ttids[iCh];
+          while (current_tower_block < trailer && iCh < numbChannels) {
+            auto const w = *current_tower_block;
+            uint8_t ttid = w & TOWER_ID_MASK;
+            uint16_t bxlocal = (w >> TOWER_BX_B) & TOWER_BX_MASK;
+            uint16_t lv1local = (w >> TOWER_L1_B) & TOWER_L1_MASK;
+            uint16_t block_length = (w >> TOWER_LENGTH_B) & TOWER_LENGTH_MASK;
+
+            // fast forward to the next good tower id (in case of recovery from an earlier header corruption)
+            while (exp_ttids[iCh] < next_tower_id) {
+              ++iCh;
+            }
+            ++iCh;
+
+            // check if the tower id in the tower header is the one expected
+            // if not try to find the next good header, point the current_tower_block to it, and extract its tower id
+            // or break if there is none
+            if (ttid != next_tower_id) {
+              next_tower_id = find_next_tower_block(current_tower_block, trailer, bx, lv1);
+              if (next_tower_id < TOWER_ID_MASK) {
+                continue;
+              } else {
+                break;
+              }
+            }
+
+            // prepare for the next iteration
+            next_tower_id = exp_ttids[iCh];
+
+            uint16_t const dccbx = bx & 0xfff;
+            uint16_t const dccl1 = lv1 & 0xfff;
+            // fov>=1 is required to support simulated data for which bx==bxlocal==0
+            if (fov >= 1 && !is_synced_towerblock(dccbx, bxlocal, dccl1, lv1local)) {
+              current_tower_block += block_length;
+              continue;
+            }
+
+            // go through all the channels
+            // get the next channel coordinates
+            uint32_t nchannels = (block_length - 1) / 3;
+
+            bool bad_block = false;
+            auto& ch_with_bad_block = alpaka::declareSharedVar<uint32_t, __COUNTER__>(acc);
+            // 1 threads per channel in this block
+            for (uint32_t ich = 0; ich < nchannels;
+                 ich += alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u]) {
```

Comment:
I guess it can happen if `nchannels` is not a multiple of `alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u]` ?

By the way, is it intended to ignore the `Elems` part of the block size ?
On CPU `alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u]` is always 1.

Reply:
The code as it is works fine with the nvidia and serial backends and gives perfect agreement with the legacy unpacker in both cases. `nchannels` is determined within the kernel by looking at the raw data and I think it can have values other than multiples of the block size.

Using `alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u]` instead leads to mismatches between legacy and the serial backend where much fewer digis are being produced. Not really sure why this happens.

Reply:
I'm only concerned (given the wording in CUDA documentation on `__syncthreads()`) if in some cases the `syncBlockThreads()` calls here could lead to problems. (and we have seen problems from `__syncthreads()` in conditional code)